Added Python Documentation (#128)
spokestack/spokestack-website: content/docs/Python/tts.md

---
title: Text-to-speech for Python
navId: TTS (Python)
description: Documentation for Spokestack's Python TTS module.
draft: false
---

Text-to-speech is a broad topic, but as far as Spokestack is concerned, there are two things your app has to handle: sending text, SSML, or Speech Markdown to be synthesized; and playing the resulting audio for your users. This guide will cover both.

## Generating Audio

The best way to synthesize speech in Spokestack is to use the `TextToSpeechManager` module. Keep in mind that this module operates independently of the `SpeechPipeline`.

`TextToSpeechManager` is initialized as follows:

```python
from spokestack.tts.manager import TextToSpeechManager
from spokestack.tts.clients.spokestack import TextToSpeechClient
from spokestack.io.pyaudio import PyAudioOutput

manager = TextToSpeechManager(
    TextToSpeechClient("spokestack_id", "spokestack_secret"), PyAudioOutput()
)

manager.synthesize(utterance="welcome to spokestack", mode="text", voice="demo-male")
```

> 

Reviewer: Include an example using SSML + the URL option please

Author: I added Speech Markdown as well. The URL option is only available when using `TextToSpeechClient` alone, but I'm not sure if someone would want automatic playback and URL at the same time. If that's useful, I can update `TextToSpeechManager` to include a synthesize to URL option also.

Reviewer: Ah, ok, so then just provide an example of using `TextToSpeechClient` alone, since you're right, it probably won't be a use case to get both the URL and the playback. Or update `TextToSpeechManager` as you suggest, whichever you think is the best way to accommodate a URL-only use case.

Author: The example is at the bottom of tts.md

---
title: Text-to-speech for Python
navId: TTS (Python)
description: Documentation for Spokestack's Python TTS module.
draft: false
---

> Text-to-speech is a broad topic, but as far as Spokestack is concerned, there are two things your app has to handle: sending text or SSML to be synthesized, and playing the resulting audio for your users. This guide will cover both.

Reviewer: Does the python tts client support SpeechMarkup?

> Text-to-speech is a broad topic, but as far as Spokestack is concerned, there are two things your app has to handle: sending text or SSML to be synthesized, and playing the resulting audio for your users. This guide will cover both.

Reviewer 2: ```suggestion
Text-to-speech is a broad topic, but as far as Spokestack is concerned, there are two things your app has to handle: sending text, SSML, or Speech Markdown to be synthesized; and playing the resulting audio for your users. This guide will cover both.
```

## Generating Audio

> The best way to synthesize speech in Spokestack is to use the TextToSpeechManager module. Keep in mind that these modules operate independently of the `SpeechPipeline`.

Reviewer 2: ```suggestion
The best way to synthesize speech in Spokestack is to use the `TextToSpeechManager` module. Keep in mind that this module operates independently of the `SpeechPipeline`.
```

> The `TextToSpeechManager` is initialized as follows:

Reviewer 2: ```suggestion
`TextToSpeechManager` is initialized as follows:
```

```python
from spokestack.tts.manager import TextToSpeechManager
from spokestack.tts.clients.spokestack import TextToSpeechClient
from spokestack.io.pyaudio import PyAudioOutput

manager = TextToSpeechManager(
    TextToSpeechClient("spokestack_id", "spokestack_secret"), PyAudioOutput()
)

> manager.synthesize(utterance="welcome to spokestack", mode="text", voice="demo-male")

Reviewer: It would be good to show an example using SSML as well.

```

> If you haven't already, you will need to [create an account](/create) or [sign in](/login) to get your API credentials. See the [TTS concept guide](/docs/Concepts/tts) for additional information about the types that are supported for the `mode` argument.

Reviewer 2: ```suggestion
If you haven't already, you will need to [create an account](/create) or [sign in](/login) to get your API credentials. See the [TTS concept guide](/docs/concepts/tts) for additional information about the types that are supported for the `mode` argument.
```

> ### Callbacks

Reviewer 2: Again, this section would be appropriate if you had callbacks specific to TTS—things like [`TTSEvent.Type`](https://www.javadoc.io/doc/io.spokestack/spokestack-android/latest/io/spokestack/spokestack/tts/TTSEvent.Type.html) in Android.

Crucially, you _don't_ usually want to hook TTS up to the pipeline like this example does, as it's skipping the NLU piece, preventing the app from doing anything with the transcript before responding.

> Similar to the other callbacks, TTS will need to be registered as a response to an event when used in the `SpeechPipeline`. Below, is a sample that speaks back what the ASR component recognized.

Reviewer: Does the TTS component have a callback for the synthesis result?

Author: No. It's only accessible via a pipeline event handler.

)
```

There are three different modes for TTS: `text`, `ssml`, `markdown`. We will go over each mode briefly here. However, if you would like a more detailed view check out the [TTS concept guide](docs/concepts/tts).

### Text

The `text` mode is for plain text without any additional markup. To synthesize plain text you do the following:

```python
manager.synthesize(utterance="welcome to spokestack", mode="text", voice="demo-male")
```

### SSML

[SSML](https://www.w3.org/TR/speech-synthesis11/) is based on XML and gives you enhanced control over pronunciation. Check out the [guide](docs/concepts/tts) for more details. You can synthesize speech from SSML like this:

```python
manager.synthesize(
    utterance="<speak>welcome to spokestack</speak>", mode="text", voice="demo-male"
)
```

### Speech Markdown

Speech Markdown is a wrapper around SSML syntax that gives some additional features as explained in the [guide](docs/concepts/tts). An example of Speech Markdown looks like this:

```python
manager.synthesize(
    utterance="See all our products at (www)[characters] dot my company dot com.",
    mode="text",
    voice="demo-male",
)
```

## Additional Synthesis Options

> If automatic playback is **not** what you are looking for we offer another option. An instance of

Reviewer 2: ```suggestion
If automatic playback is **not** what you are looking for, we offer another option. An instance of
```
